drivers/flash/disk_flash.c

```diff
@@ -0,0 +1,344 @@
+/*
+ * Copyright (c) 2023 Endress+Hauser AG
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_disk_flash
+
+#include <string.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/device.h>
+#include <zephyr/storage/disk_access.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/drivers/disk.h>
+
+LOG_MODULE_REGISTER(disk_flash, CONFIG_FLASH_LOG_LEVEL);
+
+struct disk_flash_data {
+	struct k_sem sem;
+	uint8_t *const sector_buf;
+	uint16_t const sector_buf_size;
+	uint32_t disk_total_sector_cnt;
+	uint32_t disk_sector_size;
+};
+
+struct disk_flash_config {
+	uint32_t flash_size;
+	struct flash_parameters flash_parameters;
+	char *disk_name;
+	uint32_t disk_offset;
+
+	IF_ENABLED(CONFIG_FLASH_PAGE_LAYOUT, (struct flash_pages_layout fpl;))
+};
+
+static void acquire_device(const struct device *dev)
+{
+	struct disk_flash_data *const data = dev->data;
+
+	k_sem_take(&data->sem, K_FOREVER);
+}
+
+static void release_device(const struct device *dev)
+{
+	struct disk_flash_data *const data = dev->data;
+
+	k_sem_give(&data->sem);
+}
+
+static int check_boundary(const struct device *dev, off_t offset, size_t len)
+{
+	int ret = 0;
+	const struct disk_flash_config *cfg = dev->config;
+	size_t operation_end = offset + len;
+
+	if (cfg->flash_size < operation_end || offset < 0) {
+		LOG_ERR("Operation out of bounds");
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+static int disk_flash_read(const struct device *dev, off_t offset, void *dst, size_t len)
+{
+	int ret = 0;
+	uint8_t *dst_buf = dst;
+	const struct disk_flash_config *cfg = dev->config;
+	struct disk_flash_data *data = dev->data;
+	uint32_t write_cursor = 0;
+
+	acquire_device(dev);
+
+	ret = check_boundary(dev, offset, len);
+	if (ret != 0) {
+		goto release_device;
+	}
+
+	if (data->disk_sector_size == 0) {
+		ret = -EINVAL;
+		goto release_device;
+	}
+
+	while (write_cursor < len) {
+		uint32_t read_cursor = cfg->disk_offset + offset + write_cursor;
+		uint32_t read_sector = read_cursor / data->disk_sector_size;
+		uint32_t read_padding = read_cursor % data->disk_sector_size;
+		uint32_t len_write = (len - write_cursor);
+
+		if (read_sector < 0 || read_sector >= data->disk_total_sector_cnt) {
+			ret = -EINVAL;
+			goto release_device;
+		}
```

Comment:
Shouldn't the range be tested before entering the loop?

Reply:
You are right. This check actually is not even needed anymore. With the `check_boundary` function we check that the read/write is within the defined flash partition and in `disk_flash_init` we check that the partition is within the underlying disk.